### ISSUE TITLE
canActivate=>false can result in blank screen/dead link

### ISSUE THREAD

**Issue by matthewerwin** [angular/angular#16211]
**I'm submitting a ...**  (check one with "x")
```
[ X ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

canActivate, canActivateChild will result in NavigationCancel event if 'false'/reject() happens in the Guard. Additionally the URL is reset to the base path before the navigation attempt (as though to leave the user on the page they navigated one).

This current behavior means that if you just go directly to a location in your browser (e.g. localhost:4200/some-guarded-route) and it returns false the URL will read "localhost:4200/" with a blank page as navigation is canceled.

**Expected behavior**

The right behavior to me seems that the router should try to see if there are further matches in the routing table (i.e. the guard 'false' case would cover the use case scenario of #14515). By continuing down the route table the behavior would be to land on '**' route in the worst case while also enabling conditional routing.

Minimally NavigationCancel event should determine if the resetUrlToCurrentUrlTree is an already displayed route (either rendering it if necessary or doing nothing if it's already rendered).

I will see if I can issue a PR to fix this for review. Will need to determine what the side effects would be for cases where people would expect just to stay on the currently rendered page (considering it was already rendered). However, to me I think the correct behavior for a dead link should be to go down the route table and hit '**' for a 404 or something vs doing nothing when the user clicks on it and rendering nothing if the user hits the route directly before any route has been rendered.

**Minimal reproduction of the problem with instructions**

Plunker won't work for this because you would need to be able to hit the route as though it was bookmarked to reproduce.

Effectively just create project using CLI, create a route guard, return false in the methods. Then attempt to hit the route directly without loading a root non-guarded route first.

**What is the motivation / use case for changing the behavior?**

Motivation is that you won't have the user on a blank screen if they happen to hit a guarded route without meeting the conditions.

Second part is that by resolving these scenarios by continuing down the route table you would enable conditional routing for scenarios such as:

Creating routes like:
localhost:4200/:organization
localhost:4200/:user

If the route guard found a valid organization it could allow the request to move forward vs user. This is all gravy. A start is to not give the user a blank page when resetting the route via resetUrlToCurrentUrlTree when the current Url wasn't rendered in the first place.

**Please tell us about your environment:**

Mac OS Sierra (10.12.4), WebStorm, NPM, ng serve

* **Angular version:** 2.0.X

Angular 4.0.2

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

Chrome 57.0.2987.133 (64-bit)
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript 2.2.2

* **Node (for AoT issues):** `node --version` =   
Node 7.9

**Comment by gedclack** [comment]
Yep, this also happens to me.
for temporary solution, I added `router.navigate(['somewhere'])` before returning false

**Comment by xdom** [comment]
@matthewerwin How would you handle a use-case when you actually need to cancel routing after `canActivate` returned `false`? I agree with your proposal but it would be good if both "modes" could be preserved somehow, wouldn't it? E.g. _(only my quick idea)_ return a `Promise` which:
- `resolve(true)` when `canActivate === true`
- `resolve(false)` when `canActivate === false` and should proceed (find next suitable route)
- `reject()` when `canActivate === false` and routing should be cancelled.

**Comment by matthewerwin** [comment]
This also resolves https://github.com/angular/angular/issues/14515

**Comment by andrui** [comment]
#18975

**Comment by code-tree** [comment]
I'm surprised more people haven't come across this yet... Seems to be a pretty big flaw if there's no decent solution.

Is it possible to detect if this is the first route being navigated to (i.e. on page load)? Could then redirect if so, else simply return false.

**Comment by code-tree** [comment]
This seems to be working for me:

```typescript
if (this.router.routerState.snapshot.url === '')
    this.router.navigate(['/'])
```

I assume `url` will only ever be an empty string on first load, as index gives value `'/'`

**Comment by AngularTx** [comment]
Hi , 

I too have the same issue . When canactivate returns false, the route doesn't stay with the previous one, instead goes blank ..like 

http://localhost:1000/a1
when i try to navigate to a2 - canactivate is triggered, which returns false 
and i get http:/localhost:1000 - blank  .

I my case i dont want to return to the "/" . I just need to stay on the last route navigated . ie http://localhost:1000/a1(with no refresh)

**Comment by montella1507** [comment]
I really didn't get why is the return type of canActivate   boolean, observable<boolean>

and not - for example   boolean | RouterCommands   or observable<boolean | routercommands>

so you would just:
`if (...) {
 return true;
}
return ["/"];
`
to force you not to forget on new router commands..

**Comment by AngularTx** [comment]
any updates here? i am using the latest version of angular 4.

**Comment by matthewerwin** [comment]
@jasonaden what's the word? What do you need from me?

Anyone else on this thread--if you need a build for a particular version of Angular let me know and I'll get it posted to MyGet feed so you can pull in.

**Comment by AngularTx** [comment]
Hi ,

Any solution for this issue ? or will this be taken care in the next updates? 

The issues comes in when the user changes the url from the browser url and tries navigating to a page which he is not allowed to , because of the canactivate condition returning false; 

If any alternative , please suggest.

**Comment by Goodwine** [comment]
It would be great if we could return `boolean|string|Promise<boolean|string>|Observable<boolean|string>` where if the value is a boolean then the current behaviour of bouncing the request would be kept, however if we return a string, then the router should navigate to that route relative to the attempted-to-activate route.

for example:
if I'm trying to navigate to `some/place/123` and I return `..`, I should go to `some/place`,
but if I return `../bad` then I should go to `some/place/bad`

Or... some other behavior similar to that.

**Comment by matthewerwin** [comment]
@goodwine that should be supported within my version. Please feel free to verify by installing and checking you use case -- returning false on all routes will result in dead link, reject or observable error results in error, and support for ** default route:

npm uninstall @angular/router
npm install @angular/router --registry https://myget.org/F/snaptech/npm

**Comment by matthewerwin** [comment]
@goodwine sorry, sent that too soon. The idea of having strings doesn't make a lot of sense to me. CanActivate guards are sentinel, not dynamic route builders. if you want that to be the fallback you would add next in the route table and if it needed conditionality just create yet another CanActivate.

In essence what you're asking for is supported behavior in my code but not in the way you're trying to do it which would muddy domains of responsibility

**Comment by jasonaden** [comment]
This might be something fixed with a recent update in #20803. Can you confirm?

**Comment by matthewerwin** [comment]
@jasonaden no, that one does not fix the issue and doesn't allow fall through behavior on the routing table

**Comment by AngularTx** [comment]
I am now using the latest Angular version 5+ . Seems like the issue is not solved there as well. Any idea when this will be taken care of ?

**Comment by micsel** [comment]
Hi all, try below code...

    import { Injectable } from '@angular/core';
    import { CanActivate, ActivatedRouteSnapshot, RouterStateSnapshot, Router } from '@angular/router';
    import { Observable } from 'rxjs/Observable';
    import { AuthService } from './auth.service';

    @Injectable()
    export class AuthGuard implements CanActivate {
       constructor( private authService : AuthService, private router: Router) {}
 
      canActivate(
        next: ActivatedRouteSnapshot,
        state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
    
        setTimeout(()=>{ //timer trick
          if(this.authService.isUserEmailLoggedIn==false){ 
            this.router.navigate(['']); //home page, usually logged out state
          }else{
           this.router.navigate(['dashboard']); //redirect to your after loggedin page
          } 
        },250) //try with 200, but 250 worked great, 500 is slower
      return this.authService.isUserEmailLoggedIn;
      }
   }

**Comment by andrui** [comment]
@micsel sory, but this is not the solution. With the same result you may call router.navigate() inside the component after needed checks. The logic of code is "try open route" => "canActivate? no!" => "reject component activation" => "navigate where I need programatically". But we don't need to implement own guards because we have them out of box. The only problem of this very issue is that if we open link like a/b/c  and "canActivate" of "c" returns false then router rejects all snapshots of chain to be activated. Expected result: we have access to a/b

**Comment by markgoho-EDT** [comment]
I'm also seeing this with a `canLoad` guard response of false

**Comment by matthewerwin** [comment]
@markgoho -- all of the guards are fixed in the PR. @jasonaden stated that it will be slated for review and inclusion in next few weeks sprint. Routes can all be fully guarded and will flow to the first fully available route.

**Comment by matthewerwin** [comment]
For anyone that wishes to test the solution I have builds through v6.0.0-beta.4

https://www.myget.org/feed/snaptech/package/npm/@angular/router

I haven't really had problems if the versions didn't match up with the rest of the Angular build but hopefully this will be folded in and released in 5.2.5 and absolutely for v6

**Comment by markgoho-EDT** [comment]
Still seeing a blank screen on 5.2.7

**Comment by markgoho** [comment]
Still happening in 5.2.8 -- @jasonaden are we out of luck with this one? Just wait for v6?

**Comment by rafa-as** [comment]
Same here! any updates?

**Comment by markgoho-EDT** [comment]
Sad news I'm also seeing this in v6.0.0-beta-7.

**Comment by markgoho-EDT** [comment]
Still waiting on a fix for this. Nothing in Angular 6 seems to fix it.

**Comment by Brandinga** [comment]
also struggling with this. landing on a blank screen :(

**Comment by chulian1819** [comment]
seriously, for simple things like this, angular is not production ready, will have to drop angular and use something more mature, or at least one that doesn't fail at the most basic things like this one.  Having too many issues with simple things, trying to build a prototype to show to the company why we should move to angular, bur right now I have more reasons why not to move to angular and just run away from it after  so much frustration.

**Comment by matthewerwin** [comment]
@chulian1819 Did you try one of my builds and see that it fixed your problem?

I have just posted a build that is for v6.02 to: https://www.myget.org/feed/snaptech/package/npm/@angular/router. 

Sorry for the delay syncing to the latest Angular master codebase. Took a minute because I was in Europe for all of Q1 and part of Q2 focused on other things.

As an aside -- The determination of using Angular versus any other framework for a company shouldn't be based on some of these outliers in functionality but rather focused on the philosophy of the solution set you need. The platform is quite complete and easy enough to do most things. Hopefully this will be native to the Angular router soon & they will pick up this feature.

**Comment by darthmolen** [comment]
6.2 and this is still not fixed? How can Authentication and canActivate() be considered an "outlier".

**Comment by darthmolen** [comment]
@matthewerwin I just upgraded to 6.1 and verified that this is still a bug. In fact, I made a blank project with little to no functionality and replicated the issue :( It's sad really. Authentication and Verification really aren't a priority. I realize the dev team has a back log on their hands but a security use case is far from a "fringe" use case.

https://github.com/darthmolen/test-latest-angular

```npm install
   ng serve
```
Go to the app, get a blank screen. all I did was write a guard that returned back an observable false.

**Comment by matthewerwin** [comment]
@darthmolen go bump the PR on this and hopefully it'll get merged here in the next sprint or two (https://github.com/angular/angular/pull/16416). You can also verify that the fix on the PR works for you as follows:


`
npm install @angular/router@6.1.0 --registry https://myget.org/F/snaptech/npm --save
`

**Comment by darthmolen** [comment]
@matthewerwin maybe check my work? I pushed some syntax error fixes and my test solution builds. package-lock.json is pointing to your fix for angular/router. guard always returns false, but it's not dropping to the ** route andcomponent and still showing a blank screen. Is there something else I need to do after the guard comes back?

**Comment by matthewerwin** [comment]
@darthmolen problem with NPM is that it won't actually use my registry so you're probably still getting the Angular published version of the router. I'd suggest running the npm command then checking if the code is correct by looking at node_modules/@angular/router/esm2015/src/router.js -- if you can find the word "stopSearching" then it has the correct build installed.

Yarn works much better at always pulling from the correct repository as you can set a yarnrc file like so and it'll check my registry first (which only has the router) before going to the main npm packages:

`"@angular:registry" "https://www.myget.org/F/snaptech/npm" `

**Comment by darthmolen** [comment]
even the .yarnrc didn't work. Finally ran the following command (with the .yarnrc in place)

```
yarn add @angular/router@6.1.0 --registry https://myget.org/F/snaptech/npm --save
```
I got a truthy test (your fix worked)

![image](https://user-images.githubusercontent.com/533340/44040686-7f7657d4-9ee1-11e8-8ad0-b10a43ee0ceb.png)

Now it did fall down to the ** route.

**Comment by scartoalex** [comment]
@Injectable()

export class AuthGuard implements CanActivate {

auth: any = {};

constructor(private authService: AuthService, private router: Router) {

}

canActivate() {
    if (/*user is logged in*/) {
        this.router.navigate(['/dashboard']);
        return true;
    }
    else {
        this.router.navigate(['/Login']);
    }
    return false;
}
}

**Comment by onderceylan** [comment]
I'm facing the same issue, is there anything I can help with to see the fix PR merged?

**Comment by zentoaku** [comment]
6.1.4 still the same issue

**Comment by ggjersund** [comment]
Facing the issue as well with 6.2.5 when use a `canActivateChild: [PermissionGuard]` for nested routes. I have a wildcard in the root routing module, but I am hoping to dont repeat this for each child routing module.

**Comment by namdien177** [comment]
Still experiencing this issue on v7.0.3.
Most effective way is to redirect before really returning false (which It feels pretty unlike theoretical)

**Comment by jhonland** [comment]
Having this issue on v7.0.3
I trying to achieve the following guards in a lazy loaded module

```
export const routes: Routes = [
  { path: '', component: ShopComponent },
  { path: ':slug', component: ListingComponent, canActivate: [IsCategoryGuard] },
  { path: ':slug', component: ListingComponent, canActivate: [IsProductGuard] },
];
```
if IsCategoryGuard return false, then the router navigation stops and never pass through IsProductGuard

**Comment by jhonland** [comment]
I just solve my issue in my previous comment, it end up like this
```
export const routes: Routes = [
  { path: '', component: ShopComponent },
  { path: ':slug', component: ListingComponent, canActivate: [UrlExistGuard] },
];
```
where UrlExistGuard checks wether :slug is a valid category or a valid product slug,
and if not a valid url from within UrlExistGuard redirect the user to a notFoundComponent

**Comment by cgosiak** [comment]
This is still an issue; it doesn't seem that there are really any solid workarounds. Any updates on a fix?

**Comment by matthewerwin** [comment]
@cgosiak go upvote the PR (https://github.com/angular/angular/pull/16416). 

@namdien177 if your routing module has to be aware of where to go next (if guard fails) it defeats the entire purpose of a router. Further, it makes it impossible to have independent modules that are lazy loaded.

@jhonland that doesn't allow you to have a product module and a category module & further requires them to be cross-aware. The redirect also has a bunch of other problems to it.

My request is that everyone upvote the PR (thumbs up, etc) so we can get some prioritization from Google team. Because this is a medium risk (it effects a core part of Angular) they have to be cautious & that caution causes it to keep getting pushed off for merge for other things with higher priority and lower support risk. Go upvote the PR.

8.0.0 is about to release so this is a perfect time to get this merged.

FYI @onderceylan @zentoaku

**Comment by daniel-nagy** [comment]
An ugly workaround is to mutate your route config at runtime. It even works with AOT 😂.

For example, say you wanted to serve you marketing site and your user's home page at the root of your domain.

```typescript
const appRoutes: Routes = [{
  path: '',
  children: [
    {
      path: '',
      canActivate: [MarketingPageGuard],
      loadChildren: './pages/marketing/marketing-page.module#MarketingPageModule'
    },
    {
      path: '',
      canActivate: [AuthGuard],
      loadChildren: './pages/home/home-page.module#HomePageModule'
    }
  ]
}];
```
```typescript
// canActivate method of MarketingPageGuard
canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): boolean {
  const canActivate = !this._session.isActive;

  if (!canActivate) {
    // @ts-ignore
    route.parent.routeConfig.children.reverse();
    this._router.navigate(['/']);
  }

  return canActivate;
}
```

**Comment by lucaslar** [comment]
I faced the same problem in Angular 8.0.3 and found a one-liner-solution that at least works for me: `this.router.navigate(<any[]><unknown>'');`

### For the impatient: 

```typescript
export class MyGuard implements CanActivate {

  constructor(private router: Router) {
  }

  canActivate(): boolean {

    const isConditionFulfilled = ... // my condition

    if (!isConditionFulfilled) {

      // notify user:
      console.log('You shall not pass!');

      // My magic line:
      this.router.navigate(<any[]><unknown>'');
    }
    return isConditionFulfilled;
  }
}
```

### Further information:

#### My expectations/what I wanted to achieve in my application:
Component `ConditionBasedComponent` is protected by a Guard as it can only be activated if a certain condition is fulfilled. Otherwise, the user will get informed and ...
- if called by navigating through the application: remain on page.
- if called by URL: handle like: `{path: '**', redirectTo: ''}`.

In other words, I did neither expect nor want the user to be redirected in any case.

#### The relevant part of my Routing module:
```typescript
const routes: Routes = [
  // ...
  {path: 'condition-based-component', component: ConditionBasedComponent, canActivate: [MyGuard]},
  {path: '', component: WelcomeComponent},
  {path: '**', redirectTo: ''}
];
```

#### Additional information: 
This also worked with `canActivate()` returninig an `Observable<boolean>` using `of(isConditionFulfilled)`.

**Comment by Simon-Bru** [comment]
Encountered the same issue, fixed it with the wildcard redirect too:
`
const routes: Routes = [
  // ...
  {path: '**', redirectTo: ''}
];
`

**Comment by jbojcic1** [comment]
I have route guard that returns false when some condition is met as I want user to stay on the current route if condition is not met. 
It works fine if I try to navigate to guarded route through cliking links in the app. If I enter guarded page url directly in the browser, I get redirected to 'http://localhost:4000' and even though in my routing module I have this set up:
`{ path: '', redirectTo: '/home', pathMatch: 'full' }`
I am not being redirected to home page and app stays in broken state.

**Comment by eohulbert** [comment]
> I have route guard that returns false when some condition is met as I want user to stay on the current route if condition is not met.
> It works fine if I try to navigate to guarded route through cliking links in the app. If I enter guarded page url directly in the browser, I get redirected to 'http://localhost:4000' and even though in my routing module I have this set up:
> `{ path: '', redirectTo: '/home', pathMatch: 'full' }`
> I am not being redirected to home page and app stays in broken state.

Place your AuthGuard paths in the bottom rows. I can't explain it, but it works. My guess is that once we trigger the function and it returns false and navigates /Home, we are still using the same Routes list with router-outlet instead of a new instance of the list.

Since it is the same list, it might be causing the AuthGuard rows and the rows after to be false. Yes, I know this sounds absurd, but how many times have YOU encountered errors that were completely off-base and literally due to a missing comma or something similar?

> So what's happening? Why isn't it redirecting to the exact path we told the function to navigate to?

It is, but the component isn't loading so you won't see "localhost:4200/Home" on your address bar and will just show "localhost:4200". The html will load from "src/index.html" just fine, but the component itself doesn't activate. Ultimately, defeating the purpose and ruining your day.

When I moved the AuthGuard rows to the bottom of the list, everything started to work just fine.

`const routes: Routes = [
  { path: "", redirectTo: "/Home", pathMatch: "full" },
  { path: "Home", component: HomeComponent },
  { path: "Login", component: LoginComponent },
  { path: "**", redirectTo: "/Home" },
  { path: "User", canActivate: [AuthGuard], component: UserComponent },
  { path: "Admin", canActivate: [AuthGuard], component: AdminComponent }
];`

**Comment by eohulbert** [comment]
> I have route guard that returns false when some condition is met as I want user to stay on the current route if condition is not met.
> It works fine if I try to navigate to guarded route through cliking links in the app. If I enter guarded page url directly in the browser, I get redirected to 'http://localhost:4000' and even though in my routing module I have this set up:
> `{ path: '', redirectTo: '/home', pathMatch: 'full' }`
> I am not being redirected to home page and app stays in broken state.

Place your AuthGuard paths in the bottom rows. I can't explain it, but it works. My guess is that once we trigger the function and it returns false and navigates /Home, we are still using the same Routes list with router-outlet instead of a new instance of the list.

Since it is the same list, it might be causing the AuthGuard rows and the rows after to be false. Yes, I know this sounds absurd, but how many times have YOU encountered errors that were completely off-base and literally due to a missing comma or something similar?

> So what's happening? Why isn't it redirecting to the exact path we told the function to navigate to?

It is, but the component isn't loading so you won't see "localhost:4200/Home" on your address bar and will just show "localhost:4200". The html will load from "src/index.html" just fine, but the component itself doesn't activate. Ultimately, defeating the purpose and ruining your day.

When I moved the AuthGuard rows to the bottom of the list, everything started to work just fine.

`const routes: Routes = [
  { path: "", redirectTo: "/Home", pathMatch: "full" },
  { path: "Home", component: HomeComponent },
  { path: "Login", component: LoginComponent },
  { path: "**", redirectTo: "/Home" },
  { path: "User", canActivate: [AuthGuard], component: UserComponent },
  { path: "Admin", canActivate: [AuthGuard], component: AdminComponent }
];`

**Comment by eohulbert** [comment]
> I have route guard that returns false when some condition is met as I want user to stay on the current route if condition is not met.
> It works fine if I try to navigate to guarded route through cliking links in the app. If I enter guarded page url directly in the browser, I get redirected to 'http://localhost:4000' and even though in my routing module I have this set up:
> `{ path: '', redirectTo: '/home', pathMatch: 'full' }`
> I am not being redirected to home page and app stays in broken state.

> I have route guard that returns false when some condition is met as I want user to stay on the current route if condition is not met.
> It works fine if I try to navigate to guarded route through cliking links in the app. If I enter guarded page url directly in the browser, I get redirected to 'http://localhost:4000' and even though in my routing module I have this set up:
> `{ path: '', redirectTo: '/home', pathMatch: 'full' }`
> I am not being redirected to home page and app stays in broken state.

Place your AuthGuard paths in the bottom rows. I can't explain it, but it works. My guess is that once we trigger the function and it returns false and navigates /Home, we are still using the same Routes list with router-outlet instead of a new instance of the list.

Since it is the same list, it might be causing the AuthGuard rows and the rows after to be false. Yes, I know this sounds absurd, but how many times have YOU encountered errors that were completely off-base and literally due to a missing comma or something similar?

> So what's happening? Why isn't it redirecting to the exact path we told the function to navigate to?

It is, but the component isn't loading so you won't see "localhost:4200/Home" on your address bar and will just show "localhost:4200". The html will load from "src/index.html" just fine, but the component itself doesn't activate. Ultimately, defeating the purpose and ruining your day.

When I moved the AuthGuard rows to the bottom of the list, everything started to work just fine.


```js
const routes: Routes = [
  { path: "", redirectTo: "/Home", pathMatch: "full" },
  { path: "Home", component: HomeComponent },
  { path: "Login", component: LoginComponent },
  { path: "**", redirectTo: "/Home" },
  { path: "User", canActivate: [AuthGuard], component: UserComponent },
  { path: "Admin", canActivate: [AuthGuard], component: AdminComponent }
];
```

**Comment by poke** [comment]
```
  { path: "**", redirectTo: "/Home" },
  { path: "User", canActivate: [AuthGuard], component: UserComponent },
  { path: "Admin", canActivate: [AuthGuard], component: AdminComponent }
```

That should actually prevent you from opening those routes because the catch-all route will be evaluated first. Did you check whether those still work properly?

> If I enter guarded page url directly in the browser

That's exactly what this issue is about. Accessing a guarded URL directly breaks the application.

**Comment by matthewerwin** [comment]
Further to the point is that this PR deals with having AuthGuard1 and AuthGuard2, etc...eg admin vs normal user login and having the route table operate correctly in all scenarios.

Moving to the bottom for a single guard type and no need for fall-through in a fail scenario works because it’s a simplistic case. Doesn’t matter there because if the first one fails all the others would fail as well below

**Comment by eohulbert** [comment]
> ```
>   { path: "**", redirectTo: "/Home" },
>   { path: "User", canActivate: [AuthGuard], component: UserComponent },
>   { path: "Admin", canActivate: [AuthGuard], component: AdminComponent }
> ```
> 
> 
> That should actually prevent you from opening those routes because the catch-all route will be evaluated first. Did you check whether those still work properly?
> 
> 
> > If I enter guarded page url directly in the browser
> 
> 
> That's exactly what this issue is about. Accessing a guarded URL directly breaks the application.

Your logic makes sense in that the catch all would redirect to /Home and never reach the last two paths. However, the point is to prevent users from manually entering the address in the browser to access those pages unless they are already logged in with a token. Not to break the app so they can't do anything.

The issue comes on the redirect when manually entering in the path to a page where authentication fails. It loads the page (html) but not the component. Jbojcic1 is having the issue where it loads to "localhost:4000" instead of "localhost:4000/path", as was I.

If it loads to a blank, it should redirect home, right? No. Because of base href being set to "/". Technically speaking, "localhost:4000" is really "localhost:4000/" without a path. Your catch all would be triggered on this and complete the redirect.

Possible rebuttal: "Well, you could set the navigate without the '/' and it should work."

Yes, it should. But it doesn't (and yes, this was also tested). It's not technically a blank path and if our catch all is at the end of the routes list, it's not being triggered to catch the solitary "/". When moving all of the AuthGuard to the bottom, it is now able to catch it and complete the navigation to /Home.

In production and hosting on the web, dropping the AuthGuard paths to the bottom most likely won't matter, but localhost has it's own set of problems that need to be worked around to move forward with development.

**Comment by neo** [comment]
still seeing this...

**Comment by pankajparkar** [comment]
It did worked for me, if I return `UrlTree` object

```
return myStream$.pipe(
   map(data => data.valid ? this.router.parseUrl('/') : false)
);
```

**Comment by bugger279** [comment]
```
canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): boolean {
    const url = state.url;
    this.role = localStorage.getItem('role');
    switch (url as any) {
      case '/offer':
        if (this.role === 'HRA') {
          return true;
        } else {
          this.router.navigate(['./login']);
          return false;
        }
        break;
       case '':
        if (this.role === '') {
          this.router.navigate(['./login']);
          return false;
        }
    }
```

My Temporary hack.

**Comment by cmddavid** [comment]
@bugger279 using router.navigate() inside the guard is indeed the way to go right now. I would not call it a hack but a workaround for a bug that has been around for a long time now.

**Comment by matthewerwin** [comment]
@cmddavid trying to redirect navigating mid-route using navigate only addresses part of the issue in very basic SPA’s. I’m going to explain this in detail because it’s come up a few times in this thread and discussions elsewhere.

Consider scenario 1: You have lazy loaded modules. That means your module now has to be aware of any others to know where to switch to (I.e. trying to act like a router itself that’s knowledgeable of the other routes...) A random guard, in a module that requires special permissions to possibly even load, is now trying to make decisions about other modules that may be loaded or not loaded depending on other permissions of the user. Further every module basically is now having some kindve mini alternate routes table...which doesn’t interrupt the current route request. But queues another behind it. Now with potentially every route in the composite table (Angular route table) trigging a call to navigate() trying to be smart about where to send the user.

Scenario 2: Multiple generic variable based routes. e.g. the @handle scenario. Where @organizationName used module 1 and @userName uses module 2. Now calling navigate() in the first guard will result in an endless route loop to itself.

Scenario 3: your app has more than 1 developer. Trying to keep routing straight just went out the window along with dev ta efficiency.

Calling navigate() to try to make router level decisions within a guard is a hack. A guard isn’t a router and shouldn't be trying to act like one.

Further, the angular router continues executing against the existing mid-flight route when a guard returns false. So calling navigate() just queues another route attempt. Meanwhile the router keeps calling the next series of guards. So then you could have a route that matches successfully just below your guarded route that returns false and navigate() in its infinite wisdom about what to do just because permissions failed. Now your app will then render that 1st successful matching route down the table. Then the failed guard will override that completed navigation with a new route execution. 

You can see how this also doesn’t work at all if an app has two routes that both defer to each other when a user doesn’t have a certain permission combination for example...meanwhile a route the user would have access to render just below those two in the application’s composite routing table keeps getting overridden by redirecting guards.

So then you’re writing switches and permission checks to iterate through where you should redirect the navigation in multiple modules that all have to be aware of each other and may not have been lazy loaded. With every guard in the table potentially adding to the route-stack with a navigate call...that can just trigger the same series.

Every guard, for every potential match to a route request, in the composite routing table is executed until the router finds a suitable allowed route (or one that has no guard). On that basis alone of well-known router behavior it can be fundamentally stated that: Calling navigate() from a guard is an anti-pattern.

Doing so solves only very rudimentary small app scenarios with fixed unique route combinations, no lazy loaded modules, and no use of variable routes with the same pattern. Forget having multiple devs doing anything on a project together who don’t have awareness of other modules or understand the behavior of the router. Just a recipe for bad runtime issues for the user and for bad security practices when the support team just starts granting permissions to deal with it.

I hope that clarifies the matter substantially. I do think it’s helpful to supply workarounds to people as there are plenty of situations where a hack is sufficient.

**Comment by cmddavid** [comment]
@matthewerwin thank you for the detailed description of the impact of having to go with workarounds like this. I think all the more reason to get some real progress on this issue. Maybe a solution would be to have the ability to set a routing strategy for if the guard blocks entry to the matched route. Having the ability to go for the next matched route could be sufficient?

**Comment by matthewerwin** [comment]
@cmddavid the PR to fix it has been available for a long time. The Angular team has been slammed. You can upvote the PR here:

https://github.com/angular/angular/pull/16416#issuecomment-589124977

@IgorMinar @atscott a comprehensive overview of what all that PR addresses is outlined above in my last comment on this issue.

**Comment by lmdg** [comment]
Hi @matthewerwin, first and foremost thanks so much for this feature, sorely needed IMHO. I've been using it for a while now up to version 8, however when I updated to the 9.1 rc (or 10) I noticed that it's not working as I expected any longer since the route is getting canceled and doesn't continue onto the next viable route. I've tried with setting the `cascade` option to both `true` and `cascade` but no luck. Is there something else that needed to be updated with the new cascade option in my routes?

Thankfully version 8 is still working, but I'd like to update to the latest and greatest as it'll be closer to the final api when this hopefully lands soon. Thanks again in advance.

**Comment by matthewerwin** [comment]
@lmdg I appreciate the kind words. 

For version 10 you must explicitly use (and use the MyGet upstream feed):
`"@angular/router": "10.0.0-next.3",`

For version 9 there are a variety of builds available; I recommend using the latest one I posted (9.1.0-rc.1). You can see a list of all builds I've posted here:

https://www.myget.org/feed/snaptech/package/npm/@angular/router

As there have been no real router changes of significance by the internal team the builds will work with a majority of the various builds from Angular & this one is backwards compatible.

Please note that in order to get it to work you need to enable on v10+
`RouterModule.forRoot(YourRoutingTable, {cascade: true})`

Hope this sorts it out for you.

**Comment by misterabdul** [comment]
Any news or update on this?
I'm using `"@angular/router": "~13.2.5",` and still have the same problem when `canActivate` returning `false` value.
My current workaround is only returning `true` or `UrlTree` value.
Could this be fixed? Or could we change the documentation to not to return `false`?

**Comment by tharindu-nilakshana** [comment]
i added these in app module file

```
import { JwtHelperService,JWT_OPTIONS } from '@auth0/angular-jwt';

 providers: [
    { provide: JWT_OPTIONS, useValue: JWT_OPTIONS },
    AuthGuard,
    JwtHelperService,
  ],
```
this solved my blank screen problem
